src/config.h

```diff
@@ -0,0 +1,134 @@
+#pragma once
+
+#include <cstdint>
+#include <iostream>
+#include <map>
+#include <variant>
+
+#include "types.h"
+
+namespace bpftrace {
+
+enum class ConfigSource
+{
+  default_,
+  env_var,
+};
+
+enum class ConfigKey
+{
+  ast_max_nodes,
+  cat_bytes_max,
+  debug_output,
+  no_cpp_demangle,
+  log_size,
+  map_keys_max,
+  max_probes,
+  max_bpf_progs,
+  max_type_res_iterations,
+  perf_rb_pages,
+  stack_mode,
+  strlen,
+  str_trunc_trailer,
+  user_symbol_cache_type,
+  verify_llvm_ir,
+};
+
+struct ConfigValue
+{
+  ConfigSource source;
+  std::variant<bool, uint64_t, std::string, StackMode, UserSymbolCacheType>
+      value;
+};
+
+const std::map<std::string, StackMode> STACK_MODE_MAP = {
+  { "bpftrace", StackMode::bpftrace },
+  { "perf", StackMode::perf },
+  { "raw", StackMode::raw },
+};
+
+class Config
+{
+public:
+  explicit Config(bool has_cmd = false, bool bt_verbose = false);
+  ~Config(){};
+
+  template <typename T>
+  bool set(ConfigKey key, T val, ConfigSource source)
+  {
+    auto it = config_map_.find(key);
+    if (it == config_map_.end())
+    {
+      throw std::runtime_error("Unsupported config key");
+    }
+    if (!can_set(it->second.source, source))
+    {
+      return false;
+    }
+
+    try
+    {
+      std::get<T>(it->second.value);
+    }
+    catch (std::bad_variant_access const &ex)
+    {
+      throw std::runtime_error("Type mismatch for config key");
+    }
+
+    it->second.value = (T)val;
```

Comment:
Why is the cast necessary? Looks like it should be a noop cuz of `T val` above

Reply:
Good catch. Will remove